client/app/controllers/collection/add.js

```diff
@@ -150,15 +188,43 @@ export default Ember.Controller.extend({
 
     // Delete widget and resets state
     delete_widget_signature: ['widget_object'],
-    delete_widget: function(widget_object) {
+    delete_widget: function(widget_object, parameters) {
         this.get('widgets').removeObject(widget_object.value);
-        widget_object.value = undefined;
-        widget_object.state = ['undefined'];
+        Ember.set(widget_object, 'value', undefined);
+        Ember.set(widget_object, 'state', ['undefined']);
+    },
+
+
+    disableWidget_signature: ['widget_object'],
+    disableWidget: function(widget_object, parameters) {
+        Ember.set(widget_object, 'value.disabled', true);
+    },
+
+
+    enableWidget_signature: ['widget_object'],
+    enableWidget: function(widget_object, parameters) {
+        Ember.set(widget_object, 'value.disabled', false);
+        Ember.run.next(this, function() {
+            this.get('updateState').call(this, this.get('formActions'));
+        });
+    },
+
+
+    toggleleWidget_signature: ['widget_object'],
+    toggleleWidget: function(widget_object, parameters) {
+        if (widget_object.value.disabled === false) {
+            Ember.set(widget_object, 'value.disabled', true);
+        } else {
+            Ember.set(widget_object, 'value.disabled', false);
+        }
+        Ember.run.next(this, function() {
+            this.get('updateState').call(this, this.get('formActions'));
+        });
     },
 
 
     upload_file_signature: ['file_name', 'file_data', 'node'],
-    upload_file: async function(file_name, file_data, node) {
+    upload_file: async function(file_name, file_data, node, parameters) {
         if (typeof file_name.value === 'undefined') return;
         if (typeof file_data.value === 'undefined') return;
```

Comment:
Please remove the duplicate line `if (typeof file_data.value === 'undefined') return;`